tools/rpp/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+    process: fn(&str, &mut Context) -> Result<String, ParsingError>,
+}
+
+pub struct Context {
+    macros: HashMap<String, String>,
+    line_num: u32,
+
+    // Compiling regex is slow. Only do it once per context
+    word_re: Regex,
+    block_comment_re: Regex,
+    eol_comment_re: Regex,
+}
+
+impl Context {
+    pub fn new() -> Context {
+        Context{
+            macros: HashMap::new(),
+            line_num: 0,
+
+            word_re: Regex::new(r"\b\w+\b").unwrap(),
+            block_comment_re: Regex::new(r"/\*.*\*/").unwrap(),
+            eol_comment_re: Regex::new(r"//.*$").unwrap(),
+        }
+    }
+}
+
+const KEYWORDS: &[Keyword] = &[
+    Keyword {
+        name: "#define",
+        process: process_define,
+    },
+    Keyword {
+        name: "#include",
+        process: process_include,
+    },
+];
+
+fn process_define(value: &str, ctx: &mut Context) -> Result<String, ParsingError> {
+    let mut parts = value.split_whitespace();
+    let key = match parts.next() {
+        Some(k) => k,
+        None => return Err(ParsingError::new(
+            String::from("#define has no name"),
+            ctx.line_num)
+        ),
+    };
+
+    let val = match parts.next() {
+        Some(v) => v,
+        None => return Err(ParsingError::new(
+            format!("#define {} has no value", key),
+            ctx.line_num)
+        ),
+
+    };
+
+    ctx.macros.insert(key.to_string(), val.trim().to_string());
+    Ok(String::new())
+}
+
+fn process_include(value: &str, ctx: &mut Context) -> Result<String, ParsingError> {
+    let chars: &[_] = &['"', '<', '>'];
+    let filename = value.trim_matches(chars);
+    let tmp_line = ctx.line_num;
+    ctx.line_num = 0;
+    let out = process_file(filename, ctx);
+    ctx.line_num = tmp_line;
+
+    out
+}
+
+fn evaluate_macros(line: &str, ctx: &Context) -> String {
+    let mut last = 0;
+    let mut out = String::new();
+    for m in ctx.word_re.find_iter(line) {
+        // Push everything leading up the the matched word
+        out.push_str(&line[last..m.start()]);
+
+        // Push matched word or replacement if one was found.
+        let word = &line[m.start()..m.end()];
+
+        // Recursively try to replace word. This allows macro definitions like the
+        // following:
+        // #define A 1
+        // #define B (A + 2)
+        let out_word = match ctx.macros.get(word) {
+            Some(x) => evaluate_macros(x, ctx),
+            None => word.to_string()
+        };
+        out.push_str(out_word.as_str());
+        last = m.end();
+    }
+
+    // Push everything remaining in line after the last matched word.
+    out.push_str(&line[last..]);
+
+    return out;
+}
+
+fn process_line(line: &str, ctx: &mut Context) -> Result<String, ParsingError> {
+    // Trim whitespace
+    let trimmed = line.trim();
+
+    if trimmed.starts_with("#") {
+        // For lines with keywords, remove comments
+        let replaced = ctx.eol_comment_re.replace_all(trimmed, "");
```

Comment:
Strip comments before checking if the line starts with a '#'.

Reply:
Fixed this (just stripped all block comments before doing any parsing), although now the line numbers will be a little off for error reporting. Oh well, I added a todo, it's a problem for another day.